proposed api changes for v3.7.0 (#750)
freyacodes/Lavalink: IMPLEMENTATION.md

* You must be able to send messages via a shard's mainWS connection.
* You must be able to intercept voice server updates from mainWS on your shard connection.

## Significant changes v3.5 -> v3.6
* Deprecation of all endpoints and moved them to `/v3` namespace.

> * Deprecation of all outgoing messages (play, stop, pause, seek, volume, filters, destroy, voiceUpdate & configureResuming).

Reviewer: ```suggestion
* Deprecation of all client messages (play, stop, pause, seek, volume, filters, destroy, voiceUpdate & configureResuming).
```

* You must be able to send messages via a shard's mainWS connection.
* You must be able to intercept voice server updates from mainWS on your shard connection.

## Significant changes v3.6.0 -> v3.7.0
* Moved HTTP endpoints under the new `/v3` path.
* Deprecation of the old HTTP paths.
* WebSocket handshakes should be done with `/v3/websocket`. Handshakes on `/` is now deprecated.
* Deprecation of all client-to-server messages (play, stop, pause, seek, volume, filters, destroy, voiceUpdate & configureResuming).

> * Addition of REST endpoints intended to replace client requests

Reviewer 2: ```suggestion
* Addition of REST endpoints intended to replace client requests.
```

* You must be able to send messages via a shard's mainWS connection.
* You must be able to intercept voice server updates from mainWS on your shard connection.

## Significant changes v3.6.0 -> v3.7.0
* Moved HTTP endpoints under the new `/v3` path with `/version` as the only exception.
* Deprecation of the old HTTP paths.

> * WebSocket handshakes should be done with `/v3/websocket`. Handshakes on `/` is now deprecated.

Reviewer 3: Handshakes on `/` is - uses plural and singular, either
- Handshakes on `/` are
- Handshaking on `/` is

* You must be able to send messages via a shard's mainWS connection.
* You must be able to intercept voice server updates from mainWS on your shard connection.

## Significant changes v3.5 -> v3.6
* Deprecation of all endpoints and moved them to `/v3` namespace.

> * Deprecation of all outgoing messages(play, stop, pause, seek, volume, filters, destroy, voiceUpdate & configureResuming).

Reviewer 4: ```suggestion
* Deprecation of all outgoing messages (play, stop, pause, seek, volume, filters, destroy, voiceUpdate & configureResuming).
```

* Deprecation of [`/loadtracks` response](#track-loading-api).
* Deprecation of [track decoding response](#track-decoding-api).

> * Add of new incoming message [Ready OP](#ready-op) to get `sessionId` and `resume` status.

Reviewer 4: ```suggestion
* Addition of new WebSocket dispatch [Ready OP](#ready-op) to get `sessionId` and `resume` status.
```

* `isConnectedReq`
* `sendWS`

With Lavalink 1.x the server had the responsibility of handling Discord `VOICE_SERVER_UPDATE`s as well as its own internal ratelimiting.
This remote handling makes things unnecessarily complicated and adds a lot og points where things could go wrong. 

> One problem we noticed is that since JD-AA is unaware of ratelimits on the bots gateway connection, it would keep adding

Reviewer: JD-AA?

Client-Name: The name of your client. Optionally in the format NAME/VERSION
```

> ### Incoming messages

Reviewer 4: This should perhaps be a header encompassing the aforementioned op messages to better distinguish between implementation sections.

See [LavalinkSocket.java](https://github.com/freyacodes/lavalink-client/blob/master/src/main/java/lavalink/client/io/LavalinkSocket.java) for client implementation

### Ready OP

> Tells you if you resumed successfully and your sessionId used for REST requests.

Reviewer 4: ```suggestion
Dispatched by Lavalink upon successful connection and authorization. Contains fields determining if resuming was successful, as well as the session ID.
```

| karaoke?    | [Karaoke](#karaoke)    object      | Lets you eliminate part of a band, usually targeting vocals.                                      |
| timescale?  | [Timescale](#timescale) object     | Lets you change the speed, pitch, and rate                                                        |
| tremolo?    | [Tremolo](#tremolo) object         | Lets you create a shuddering effect, where the volume quickly oscillates                          |
| vibrato?    | [Vibrato](#vibrato) object         | Lets you create a shuddering effect, where the pitch quickly oscillates                           |
| rotation?   | [Rotation](#rotation) object       | Lets you rotate the sound around the stereo channels/user headphones aka Audio Panning            |
| distortion? | [Distortion](#distortion) object   | Lets you distort the audio                                                                        |
| channelMix? | [Channel Mix](#channel-mix) object | Lets you mix both channels (left and right)                                                       |

> | lowPass?    | [Low Pass](#low-pass)object        | Lets you filter higher frequencies                                                                |

Reviewer 2: ```suggestion
| lowPass?    | [Low Pass](#low-pass) object        | Lets you filter higher frequencies                                                                |
```

##### Equalizer
There are 15 bands (0-14) that can be changed.
"gain" is the multiplier for the given band. The default value is 0. Valid values range from -0.25 to 1.0,
where -0.25 means the given band is completely muted, and 0.25 means it is doubled. Modifying the gain could also change the volume of the output.

| Field | Type  | Description            |
|-------|-------|------------------------|

> | bands | int   | The band(0 to 14)      |
> | gain  | float | The gain(-0.25 to 1.0) |

Reviewer 2: ```suggestion
| bands | int   | The band (0 to 14)      |
| gain  | float | The gain (-0.25 to 1.0) |
```

    "cosScale": 1.0,
    "tanOffset": 0.0,
    "tanScale": 1.0,
    "offset": 0.0,
    "scale": 1.0
  },
  "channelMix": {
    "leftToLeft": 1.0,
    "leftToRight": 0.0,
    "rightToLeft": 0.0,
    "rightToRight": 1.0
  },
  "lowPass": {
    "smoothing": 20.0
  }
}
```
</details>

---

#### Destroy Player
Destroys the player for this guild in this session.
```
DELETE /v3/sessions/{sessionId}/players/{guildId}
```

Response: 

204 - No Content

---

#### Update Session
Updates the session with a resuming key and timeout.
```
PATCH /v3/sessions/{sessionId}
```

Request:

| Field       | Type   | Description                                              |
|-------------|--------|----------------------------------------------------------|
| resumingKey | string | The resuming key to be able to resume this session later |
| timeout     | int    | The timeout in seconds                                   |

<details>
<summary>Example Payload</summary>

```json
{
  "resumingKey": "...",
  "timeout": 0
}
```
</details>

Response:

204 - No Content

---

#### Track Loading
This endpoint is used to resolve audio tracks for use with the [Update Player](#update-player) endpoint.

> > `/loadtracks?identifier=dQw4w9WgXcQ` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/loadtracks?identifier=dQw4w9WgXcQ` is deprecated and marked for removal in v4
```

Depending on your Discord library, it may be possible to take advantage of the library's OP 4 handling. For instance,
the JDA client takes advantage of JDA's websocket write thread to send OP 4s for connects, disconnects and reconnects.

</details>

## Protocol
### Opening a connection
You can establish a WebSocket connection against the path `/v3/websocket`

When opening a websocket connection, you must supply 3 required headers:
```
Authorization: Password matching the server config

> User-Id: The user id of the bot you are playing music with

Reviewer 2: ```suggestion
User-Id: The user id of the bot you are playing music with, in string format
```
Not sure if you want this change since this line is not being touched in this PR, but it may be good to let people know to not send an int

Author: headers are always strings so this is not needed

Client-Name: The name of your client. Optionally in the format NAME/VERSION
```

Optionally you can supply the `Resume-Key` header. For more information on resuming see [Resuming](#resuming-lavalink-sessions)

### Websocket Messages
Fields marked with `?` are optional.
Types marked with `?` are nullable.

Websocket messages all follow the following standard format:

| Field | Type                 | Description                           |
|-------|----------------------|---------------------------------------|
| op    | [OP Type](#op-types) | The op type                           |
| ...   | ...                  | Extra fields depending on the op type |

<details>
<summary>Example Payload</summary>

```yaml
{
  "op": "...",
  ...
}
```
</details>

#### OP Types

| OP Type                           | Description                                                  |
|-----------------------------------|--------------------------------------------------------------|

> | [ready](#ready-op)                | Emitted when you successfully connected to the Lavalink node |

Reviewer 2: ```suggestion
| [ready](#ready-op)                | Emitted when you successfully connect to the Lavalink node   |
```

> | [playerUpdate](#player-update-op) | Emitted when every x seconds with the latest player state    |

Reviewer 2: ```suggestion
| [playerUpdate](#player-update-op) | Emitted every x seconds with the latest player state         |
```

| [stats](#stats-op)                | Emitted when the node sends stats once per minute            |
| [event](#event-op)                | Emitted when a player or voice event is emitted              |

#### Ready OP
Dispatched by Lavalink upon successful connection and authorization. Contains fields determining if resuming was successful, as well as the session ID.

| Field     | Type   | Description                                                                                    |
|-----------|--------|------------------------------------------------------------------------------------------------|
| resumed?  | bool   | If the session is resumed. (Only present if a session id was present when connecting)          |
| sessionId | string | The Lavalink session ID of this connection. Not to be confused with a Discord voice session id |

<details>
<summary>Example Payload</summary>

```json
{
  "op": "ready",
  "resumed": false,
  "sessionId": "..."
}
```
</details>

---

#### Player Update OP

> Dispatched every x(configurable in `application.yml`) seconds with the current state of the player.

Reviewer 2: ```suggestion
Dispatched every x (configurable in `application.yml`) seconds with the current state of the player.
```

```yaml
{
  "op": "event",
  "type": "...",
  "guildId": "...",
  ...
}
```
</details>

##### Event Types
| Event Type                                    | Description                                                              |
|-----------------------------------------------|--------------------------------------------------------------------------|
| [TrackStartEvent](#trackstartevent)           | Emitted when a track starts playing                                      |
| [TrackEndEvent](#trackendevent)               | Emitted when a track ends                                                |
| [TrackExceptionEvent](#trackexceptionevent)   | Emitted when a track throws an exception                                 |
| [TrackStuckEvent](#trackstuckevent)           | Emitted when a track gets stuck while playing                            |
| [WebSocketClosedEvent](#websocketclosedevent) | Emitted when the websocket connection to discord voice servers is closed |

##### TrackStartEvent
Emitted when a track starts playing.

| Field   | Type   | Description                              |
|---------|--------|------------------------------------------|

> | track   | string | The track that started playing           |

Reviewer 2: Should there be a description here for the track that indicates it's one of the b64 strings?

|--------|-------------------------------------|--------------------------------------|
| track  | string                              | The encoded track that ended playing |
| reason | [TrackEndReason](#track-end-reason) | The reason the track ended           |

##### Track End Reason
| Reason        | Description                | May Start Next |
|---------------|----------------------------|----------------|
| `FINISHED`    | The track finished playing | true           |
| `LOAD_FAILED` | The track failed to load   | true           |
| `STOPPED`     | The track was stopped      | false          |
| `REPLACED`    | The track was replaced     | false          |
| `CLEANUP`     | The track was cleaned up   | false          |

<details>
<summary>Example Payload</summary>

```json
{
  "op": "event",
  "type": "TrackEndEvent",
  "guildId": "...",
  "track": "...",
  "reason": "FINISHED"
}
```
</details>

---

##### TrackExceptionEvent
Emitted when a track throws an exception.

| Field     | Type                                  | Description                                |
|-----------|---------------------------------------|--------------------------------------------|

> | track     | string                                | The encoded track that threw the exception |

Reviewer 5: Could we please keep the names consistent? It would make developing libraries for it much more intuitive. 
The B64 is currently refered to in different ways (`track`, `trackData`, `encoded` and `encodedTrack`)

Emitted when a track gets stuck while playing.

| Field       | Type   | Description                                     |
|-------------|--------|-------------------------------------------------|
| track       | string | The track that got stuck                        |
| thresholdMs | int    | The threshold in milliseconds that was exceeded |

<details>
<summary>Example Payload</summary>

```json
{
  "op": "event",
  "type": "TrackStuckEvent",
  "guildId": "...",
  "track": "...",
  "thresholdMs": 123456789
}
```
</details>

---

##### WebSocketClosedEvent
Emitted when an audio web socket (to Discord) is closed.
This can happen for various reasons (normal and abnormal), e.g. when using an expired voice server update.
4xxx codes are usually bad.
See the [Discord Docs](https://discordapp.com/developers/docs/topics/opcodes-and-status-codes#voice-voice-close-event-codes).

| Field    | Type   | Description                                                                                                                       |
|----------|--------|-----------------------------------------------------------------------------------------------------------------------------------|

> | code     | int    | The [discord close event code](https://discord.com/developers/docs/topics/opcodes-and-status-codes#voice-voice-close-event-codes) |

Reviewer 2: ```suggestion
| code     | int    | The [Discord close event code](https://discord.com/developers/docs/topics/opcodes-and-status-codes#voice-voice-close-event-codes) |
```

<details>
<summary>Example Payload</summary>

```json
{
  "op": "event",
  "type": "WebSocketClosedEvent",
  "guildId": "...",
  "code": 4006,
  "reason": "Your session is no longer valid.",
  "byRemote": true
}
```
</details>

---

### REST API
Lavalink exposes a REST API to allow for easy control of the players.
Most routes require the Authorization header with the configured password.
```
Authorization: youshallnotpass
```

Routes are prefixed with `/v3` as of `v3.7.0`. Routes without an API prefix are to be removed in v4.

#### Get Players
Returns a list of players in this specific session.
```
GET /v3/sessions/{sessionId}/players
```

##### Player
| Field   | Type                               | Description                                           |
|---------|------------------------------------|-------------------------------------------------------|
| guildId | string                             | The guild id of the player                            |
| track   | ?[Track](#track) object            | The current playing track                             |

> | volume  | int                                | The volume of the player, range 0-1000, in percentage |

Author: 0 = 0%, 1000 = 1000% so yes this should stay

| paused  | bool                               | Whether the player is paused                          |
| voice   | [Voice State](#voice-state) object | The voice state of the player                         |
| filters | ?[Filters](#filters) object        | The filters used by the player                        |              

##### Track
| Field   | Type                             | Description                                                                   |
|---------|----------------------------------|-------------------------------------------------------------------------------|

> | encoded | string                           | The base64 encoded track data                                                 |

Reviewer 5: Same as my previous comment

> | Field   | Type                             | Description                                                                   |
> |---------|----------------------------------|-------------------------------------------------------------------------------|
> | encoded | string                           | The base64 encoded track data                                                 |
> | track   | string                           | The base64 encoded track data (DEPRECATED as of v3.7.0 and for removal in v4) |
> | info    | [Track Info](#track-info) object | Info about the track                                                          |

Reviewer 2: ```suggestion
| Field   | Type                             | Description                                                                          |
|---------|----------------------------------|--------------------------------------------------------------------------------------|
| encoded | string                           | The base64 encoded track data                                                        |
| track   | string                           | The base64 encoded track data (DEPRECATED as of v3.7.0 and marked for removal in v4) |
| info    | [Track Info](#track-info) object | Info about the track                                                                 |
```

|------------|--------|---------------------------------------------------------------------------------------------|
| token      | string | The Discord voice token to authenticate with                                                |
| endpoint   | string | The Discord voice endpoint to connect to                                                    |
| sessionId  | string | The Discord voice session id to authenticate with                                           |
| connected? | bool   | Whether the player is connected. Response only                                              |
| ping?      | int    | Roundtrip latency in milliseconds to the voice gateway (-1 if not connected). Response only |

`token`, `endpoint`, and `sessionId` are the 3 required values for connecting to one of Discord's voice servers.
`sessionId` is provided by the Voice State Update event sent by Discord, whereas the `endpoint` and `token` are provided
with the Voice Server Update. Please refer to https://discord.com/developers/docs/topics/gateway-events#voice

<details>
<summary>Example Payload</summary>

```yaml
[
  {
    "guildId": "...",
    "track": {

>       "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",

Reviewer 5: Same as my previous comment

Author: this is an oversight & should be updated

      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000, 
      "isStream": false,
      "position": 0, 
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    },
    "volume": 100,
    "paused": false,
    "voice": {
      "token": "...",
      "endpoint": "...",
      "sessionId": "...",
      "connected": true,
      "ping": 10 
    },
    "filters": { ... }
  }
]
```
</details>

---

#### Get Player
Returns the player for this guild in this session.
```
GET /v3/sessions/{sessionId}/players/{guildId}
```

Response:

[Player Object](#Player)
<details>
<summary>Example Payload</summary>

```yaml
{
  "guildId": "...",
  "track": {

>     "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",

Reviewer 5: Same as my previous comment

    "author": "RickAstleyVEVO",
    "length": 212000,
    "isStream": false,
    "position": 0,
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,
    "ping": 10
  },
  "filters": { ... }
}
```
</details>

---

#### Update Player
Updates the player for this guild.
```
PATCH /v3/sessions/{sessionId}/players/{guildId}?noReplace=true
```

Query Params:

| Field     | Type | Description                                                                  |
|-----------|------|------------------------------------------------------------------------------|
| noReplace | bool | Whether to replace the current track with the new track. Defaults to `false` |

Request:

##### Player Update
| Field         | Type                               | Description                                                                   |
|---------------|------------------------------------|-------------------------------------------------------------------------------|

> | encodedTrack? | ?string                            | The encoded track data to play. `null` stops the current track                |

Author: `trackData` should be removed in all cases this is just an oversight in the examples
`track` refers to the whole track including info
`encoded` refers to the track info encoded

and `encodedTrack` only has one use in the `Player Update` endpoint as it stands for `track.encoded`
simply having encoded there would be confusing

> | encodedTrack? | ?string                            | The encoded track data to play. `null` stops the current track                |

Author: we decided to rename the `track` field in the track events to `encodedTrack` to match the rest of the new api.
this should now be consistent and I hope what you wanted :)

| identifier?   | string                             | The track identifier to play                                                  |
| position?     | int                                | The position in milliseconds                                                  |
| endTime?      | int                                | The end time in milliseconds                                                  |

> | volume?       | int                                | The volume from 0 to 200                                                      |

Reviewer 2: Is 0 -> 200 correct?

| paused?       | bool                               | Whether the player is paused                                                  |
| filters?      | ?[Filters](#filters) object        | The new filters to apply. `null` will reset all filters                       |
| voice?        | [Voice State](#voice-state) object | Information required for connecting to Discord, without `connected` or `ping` |

> Note that `encodedTrack` and `identifier` are mutually exclusive.

Reviewer 5: Same as my previous comment

When `identifier` is used, Lavalink will try to resolve the identifier as a single track. An error response is generated
if no tracks or a playlist is resolved.

<details>
<summary>Example Payload</summary>

```yaml
{

>   "encodedTrack": "...",

Reviewer 5: Same as my previous comment

```yaml
{
  "guildId": "...",
  "track": {
    "encoded": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "info": {
      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000,
      "isStream": false,
      "position": 0,
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    }
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,
    "ping": 10 
  },
  "filters": { ... }
}
```
</details>

---

#### Filters
Filters are used in above requests and look like this

| Field       | Type                               | Description                                                                                       |
|-------------|------------------------------------|---------------------------------------------------------------------------------------------------|

> | volume?     | float                              | Lets you adjust the player volume from 0 to 5.0 where 1.0 is 100%. Values >1.0 may cause clipping |

Reviewer 2: Is there a reason why a filter can accept volumes of 0% -> 500% where the volume op is 0% -> 1000%?

Author: honestly I have no idea where that value comes from. I can't find any limit in lavaplayer. Maybe someone else can help out here

Author: Yeah I was talking about filter, not the player volume. I found the other one myself already but ty

Author: the limit is enforced right here
https://github.com/natanbc/lavadsp/blob/master/src/main/java/com/github/natanbc/lavadsp/volume/VolumePcmAudioFilter.java
so that's your reason :)

| karaoke?    | [Karaoke](#karaoke)    object      | Lets you eliminate part of a band, usually targeting vocals                                       |
| timescale?  | [Timescale](#timescale) object     | Lets you change the speed, pitch, and rate                                                        |
| tremolo?    | [Tremolo](#tremolo) object         | Lets you create a shuddering effect, where the volume quickly oscillates                          |
| vibrato?    | [Vibrato](#vibrato) object         | Lets you create a shuddering effect, where the pitch quickly oscillates                           |
| rotation?   | [Rotation](#rotation) object       | Lets you rotate the sound around the stereo channels/user headphones aka Audio Panning            |
| distortion? | [Distortion](#distortion) object   | Lets you distort the audio                                                                        |
| channelMix? | [Channel Mix](#channel-mix) object | Lets you mix both channels (left and right)                                                       |

> | lowPass?    | [Low Pass](#low-pass) object       | Lets you filter higher frequencies                                                                |

Reviewer 2: | Field       | Type                               | Description                                                                                         |
|-------------|------------------------------------|-----------------------------------------------------------------------------------------------------|
| volume?     | float                              | Lets you adjust the player volume from 0.0 to 5.0 where 1.0 is 100%. Values >1.0 may cause clipping |
| equalizer?  | array of [Equalizers](#equalizer)  | Lets you adjust 15 different bands                                                                  |
| karaoke?    | [Karaoke](#karaoke)    object      | Lets you eliminate part of a band, usually targeting vocals                                         |
| timescale?  | [Timescale](#timescale) object     | Lets you change the speed, pitch, and rate                                                          |
| tremolo?    | [Tremolo](#tremolo) object         | Lets you create a shuddering effect, where the volume quickly oscillates                            |
| vibrato?    | [Vibrato](#vibrato) object         | Lets you create a shuddering effect, where the pitch quickly oscillates                             |
| rotation?   | [Rotation](#rotation) object       | Lets you rotate the sound around the stereo channels/user headphones aka Audio Panning              |
| distortion? | [Distortion](#distortion) object   | Lets you distort the audio                                                                          |
| channelMix? | [Channel Mix](#channel-mix) object | Lets you mix both channels (left and right)                                                         |
| lowPass?    | [Low Pass](#low-pass) object       | Lets you filter higher frequencies                                                                  |

| 1    | 40 Hz     |
| 2    | 63 Hz     |
| 3    | 100 Hz    |
| 4    | 160 Hz    |
| 5    | 250 Hz    |
| 6    | 400 Hz    |
| 7    | 630 Hz    |
| 8    | 1000 Hz   |
| 9    | 1600 Hz   |
| 10   | 2500 Hz   |
| 11   | 4000 Hz   |
| 12   | 6300 Hz   |
| 13   | 10000 Hz  |
| 14   | 16000 Hz  |

</details>

| Field | Type  | Description             |
|-------|-------|-------------------------|
| bands | int   | The band (0 to 14)      |
| gain  | float | The gain (-0.25 to 1.0) |

##### Karaoke
Uses equalization to eliminate part of a band, usually targeting vocals.

| Field        | Type  | Description                                                             |
|--------------|-------|-------------------------------------------------------------------------|
| level?       | float | The level (0 to 1.0 where 0.0 is no effect and 1.0 is full effect)      |
| monoLevel?   | float | The mono level (0 to 1.0 where 0.0 is no effect and 1.0 is full effect) |
| filterBand?  | float | The filter band                                                         |
| filterWidth? | float | The filter width                                                        |

##### Timescale

> Changes the speed, pitch, and rate. All default to 1.

Reviewer 2: ```suggestion
Changes the speed, pitch, and rate. All default to 1.0.
```

| Field  | Type  | Description              |
|--------|-------|--------------------------|
| speed? | float | The playback speed 0 ≤ x |
| pitch? | float | The pitch 0 ≤ x          |
| rate?  | float | The rate 0 ≤ x           |

##### Tremolo
Uses amplification to create a shuddering effect, where the volume quickly oscillates.
https://en.wikipedia.org/wiki/File:Fuse_Electronics_Tremolo_MK-III_Quick_Demo.ogv

| Field      | Type  | Description                  |
|------------|-------|------------------------------|
| frequency? | float | The frequency 0 < x          |
| depth?     | float | The tremolo depth 0 < x ≤ 1  |

##### Vibrato
Similar to tremolo. While tremolo oscillates the volume, vibrato oscillates the pitch.

| Field      | Type  | Description                 |
|------------|-------|-----------------------------|
| frequency? | float | The frequency 0 < x ≤ 14    |

> | depth?     | float | The vibrato depth 0 < x ≤ 1 |

Reviewer 2: | Field  | Type  | Description                |
|--------|-------|----------------------------|
| speed? | float | The playback speed 0.0 ≤ x |
| pitch? | float | The pitch 0.0 ≤ x          |
| rate?  | float | The rate 0.0 ≤ x           |

##### Tremolo
Uses amplification to create a shuddering effect, where the volume quickly oscillates.
https://en.wikipedia.org/wiki/File:Fuse_Electronics_Tremolo_MK-III_Quick_Demo.ogv

| Field      | Type  | Description                      |
|------------|-------|----------------------------------|
| frequency? | float | The frequency 0.0 < x            |
| depth?     | float | The tremolo depth 0.0 < x ≤ 1.0  |

##### Vibrato
Similar to tremolo. While tremolo oscillates the volume, vibrato oscillates the pitch.

| Field      | Type  | Description                     |
|------------|-------|---------------------------------|
| frequency? | float | The frequency 0.0 < x ≤ 14.0    |
| depth?     | float | The vibrato depth 0.0 < x ≤ 1.0 |

    "smoothing": 20.0
  }
}
```
</details>

---

#### Destroy Player
Destroys the player for this guild in this session.
```
DELETE /v3/sessions/{sessionId}/players/{guildId}
```

Response: 

204 - No Content

---

#### Update Session
Updates the session with a resuming key and timeout.
```
PATCH /v3/sessions/{sessionId}
```

Request:

| Field       | Type   | Description                                              |
|-------------|--------|----------------------------------------------------------|
| resumingKey | string | The resuming key to be able to resume this session later |
| timeout     | int    | The timeout in seconds                                   |

<details>
<summary>Example Payload</summary>

```json
{
  "resumingKey": "...",
  "timeout": 0
}
```
</details>

Response:

204 - No Content

---

#### Track Loading
This endpoint is used to resolve audio tracks for use with the [Update Player](#update-player) endpoint.
> `/loadtracks?identifier=dQw4w9WgXcQ` is deprecated and marked for removal in v4
```
GET /v3/loadtracks?identifier=dQw4w9WgXcQ
```
<details>
<summary>Response(Deprecated)</summary>

```yaml
{
  "loadType": "TRACK_LOADED",
  "tracks": [
    {

>       "encoded": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",

Reviewer 5: Same as my previous comment

<details>
<summary>Example Payload</summary>

```json
{
  "op": "event",
  "type": "WebSocketClosedEvent",
  "guildId": "...",
  "code": 4006,
  "reason": "Your session is no longer valid.",
  "byRemote": true
}
```
</details>

---

### REST API
Lavalink exposes a REST API to allow for easy control of the players.
Most routes require the Authorization header with the configured password.
```
Authorization: youshallnotpass
```

Routes are prefixed with `/v3` as of `v3.7`. Routes without an API prefix are to be removed in v4.

#### Get Players
Returns a list players in this specific session.
```
GET /v3/sessions/{sessionId}/players
```

##### Player
| Field   | Type                               | Description                                          |
|---------|------------------------------------|------------------------------------------------------|
| guildId | string                             | The guild id of the player                           |
| track   | ?[Track](#track) object            | The current playing track                            |

> | volume  | int                                | The volume of the player, range 0-150, in percentage |

Author: this is actually capped by lavaplayer at 1000 https://github.com/Walkyst/lavaplayer-fork/blob/original/main/src/main/java/com/sedmelluq/discord/lavaplayer/player/DefaultAudioPlayer.java#L290 and not currently capped by lavalink at all

|---------------|------------------------------------|-------------------------------------------------------------------------------|
| encodedTrack? | ?string                            | The encoded track data to play. `null` stops the current track                |
| identifier?   | string                             | The track identifier to play                                                  |
| position?     | int                                | The position in milliseconds                                                  |
| endTime?      | int                                | The end time in milliseconds                                                  |
| volume?       | int                                | The volume from 0 to 200                                                      |
| paused?       | bool                               | Whether the player is paused                                                  |
| filters?      | ?[Filters](#filters) object        | The new filters to apply. `null` will reset all filters                       |

> | voiceSession? | [Voice State](#voice-state) object | Information required for connecting to Discord, without `connected` or `ping` |

Reviewer 3: This does not seem to match the example:
`voiceSession?`
`sessionId`
`event`

    "rightToLeft": 0.0,
    "rightToRight": 1.0
  },
  "lowPass": {
    "smoothing": 20.0
  }
}
```
</details>

---

#### Destroy Player
Destroys the player for this guild in this session.
```
DELETE /v3/sessions/{sessionId}/players/{guildId}
```

Response: 

204 - No Content

---

#### Update Session
Updates the session with a resuming key and timeout.
```
PATCH /v3/sessions/{sessionId}
```

Request:

| Field       | Type   | Description                                              |
|-------------|--------|----------------------------------------------------------|
| resumingKey | string | The resuming key to be able to resume this session later |
| timeout     | int    | The timeout in seconds                                   |

<details>
<summary>Example Payload</summary>

```json
{
  "resumingKey": "...",
  "timeout": 0
}
```
</details>

Response:

204 - No Content

---

#### Track Loading
This endpoint is used to resolve audio tracks for use with the [Update Player](#update-player) endpoint.
> `/loadtracks?identifier=dQw4w9WgXcQ` is deprecated and marked for removal in v4
```
GET /v3/loadtracks?identifier=dQw4w9WgXcQ
```
<details>
<summary>Response(Deprecated)</summary>

> ```yaml

Author: we use yaml here because of comments. see: https://github.com/freyacodes/Lavalink/pull/750#pullrequestreview-1122080700

    "paused": false,
    "voice": {
      "token": "...",
      "endpoint": "...",
      "sessionId": "...",
      "connected": true,
      "ping": 10 
    },
    "filters": { ... }
  }
]
```
</details>

---

#### Get Player
Returns the player for this guild in this session.
```
GET /v3/sessions/{sessionId}/players/{guildId}
```

Response:

[Player Object](#Player)
<details>
<summary>Example Payload</summary>

```yaml
{
  "guildId": "...",
  "track": {
    "encoded": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000,
    "isStream": false,
    "position": 0,
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,
    "ping": 10
  },
  "filters": { ... }
}
```
</details>

---

#### Update Player

> Updates the player for this guild.

Reviewer 5: One thing that wasn't immediately obvious - is what are the new ops that give equivalent behaviour to the deprecated websocket ops.

Seek, pause etc - while the fields are optional so I assume that if I send a payload with just position it will seek, just paused it will pause it may benefit users if you are explicit about this and give at least 1 example:

Say previously your send a seek op, now you make a patch call with just the position key or something like that

Author: maybe we should instead highlight this in the significant changes/deprecated op docs?

Author: with the new api you can seek, pause & set filters at the same time as an example

Author: how do you like the changes in https://github.com/freyacodes/Lavalink/commit/d1006033db73a574c5d9457cc899ff01cd348f64 ?

Request:

##### Player Update
| Field         | Type                               | Description                                                                   |
|---------------|------------------------------------|-------------------------------------------------------------------------------|
| encodedTrack? | ?string                            | The encoded track data to play. `null` stops the current track                |
| identifier?   | string                             | The track identifier to play                                                  |
| position?     | int                                | The position in milliseconds                                                  |
| endTime?      | int                                | The end time in milliseconds                                                  |
| volume?       | int                                | The volume from 0 to 1000                                                     |
| paused?       | bool                               | Whether the player is paused                                                  |

> | filters?      | ?[Filters](#filters) object        | The new filters to apply. `null` will reset all filters                       |

Reviewer 5: if  `filters=null` resets all the filters is it save to assume `{"filters": {"timescale": Null}}` will reset individual filters? unless I missed it nowhere in this doc we specify that.

Author: in the current api the whole filters get overwritten by the `filters`. This is the case here too
just omit it and it will get overwritten. so in this case `filters: {}` acts the same as `filters: null`

Author: I updated the docs on this and it should be more clearer now

|------------|---------|-------------------------------------------|
| identifier | string  | The track identifier                      |
| isSeekable | bool    | Whether the track is seekable             |
| author     | string  | The track author                          |
| length     | int     | The track length in milliseconds          |
| isStream   | bool    | Whether the track is a stream             |
| position   | int     | The track position in milliseconds        |
| title      | string  | The track title                           |
| uri        | ?string | The track uri                             |
| sourceName | string  | The track source name                     |

##### Voice State
| Field      | Type   | Description                                                                                 |
|------------|--------|---------------------------------------------------------------------------------------------|
| token      | string | The Discord voice token to authenticate with                                                |
| endpoint   | string | The Discord voice endpoint to connect to                                                    |

> | sessionId  | string | The Discord voice session id to authenticate with                                           |

Reviewer 5: The API is currently returning `sessionID` instead of `sessionId`

      "token": "...",
      "endpoint": "...",
      "sessionId": "...",
      "connected": true,
      "ping": 10
    },
    "filters": { ... }
  }
]
```

</details>

---

#### Get Player

Returns the player for this guild in this session.

```
GET /v3/sessions/{sessionId}/players/{guildId}
```

Response:

[Player Object](#Player)
<details>
<summary>Example Payload</summary>

```yaml
{
  "guildId": "...",
  "track": {
    "encoded": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000,
    "isStream": false,
    "position": 0,
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,
    "ping": 10
  },
  "filters": { ... }
}
```

</details>

---

#### Update Player

Updates the player for this guild.

```

> PATCH /v3/sessions/{sessionId}/players/{guildId}?noReplace=true

Reviewer 5: For the benefit of being able to differ from creation and update in the future I think going the PUT route now would avoid unnecessary breakage later even if they are essentially the same now in the future we may want to differ between an update and creation.

> PATCH /v3/sessions/{sessionId}/players/{guildId}?noReplace=true

Author: creation is noted

  "channelMix": {
    "leftToLeft": 1.0,
    "leftToRight": 0.0,
    "rightToLeft": 0.0,
    "rightToRight": 1.0,
  }

                  // Higher frequencies get suppressed, while lower frequencies pass through this filter, thus the name low pass.
  // Any smoothing values equal to, or less than 1.0 will disable the filter.
  "lowPass": {
    "smoothing": 20.0 // 1.0 < x
  }
}
```

#### Destroy a player (DEPRECATED)

> The `destroy` op is deprecated and marked for removal in v4, use [Destroy Player Endpoint](#destroy-player) instead.

Tell the server to potentially disconnect from the voice server and potentially remove the player with all its data.
This is useful if you want to move to a new node for a voice connection. Calling this op does not affect voice state,
and you can send the same VOICE_SERVER_UPDATE to a new node.

```json
{
  "op": "destroy",
  "guildId": "..."
}
```

</details>

---

# Common pitfalls

Admittedly Lavalink isn't inherently the most intuitive thing ever, and people tend to run into the same mistakes over again. Please double-check the following if you run into problems developing your client, and you can't connect to a voice channel or play audio:

> 1. Check that you are intercepting **VOICE_SERVER_UPDATE**s to **Lavalink**. Do not edit the event object from Discord.

Reviewer 3: It is no longer an "event object".

  "channelMix": {
    "leftToLeft": 1.0,
    "leftToRight": 0.0,
    "rightToLeft": 0.0,
    "rightToRight": 1.0,
  }

                  // Higher frequencies get suppressed, while lower frequencies pass through this filter, thus the name low pass.
  // Any smoothing values equal to, or less than 1.0 will disable the filter.
  "lowPass": {
    "smoothing": 20.0 // 1.0 < x
  }
}
```

#### Destroy a player (DEPRECATED)

> The `destroy` op is deprecated and marked for removal in v4, use [Destroy Player Endpoint](#destroy-player) instead.

Tell the server to potentially disconnect from the voice server and potentially remove the player with all its data.
This is useful if you want to move to a new node for a voice connection. Calling this op does not affect voice state,
and you can send the same VOICE_SERVER_UPDATE to a new node.

```json
{
  "op": "destroy",
  "guildId": "..."
}
```

</details>

---

# Common pitfalls

Admittedly Lavalink isn't inherently the most intuitive thing ever, and people tend to run into the same mistakes over again. Please double-check the following if you run into problems developing your client, and you can't connect to a voice channel or play audio:

> 1. Check that you are intercepting **VOICE_SERVER_UPDATE** and **VOICE_STATE_UPDATES**s to **Lavalink**. You only need the `endpoint`, `token` and `session_id`.

Author: the session id comes from the voice state update

}
```

### Get Players API
Returns all players in this specific session.
```
GET /v3/sessions/{sessionId}/players
Authorization: youshallnotpass
```

Response:
```json
[
  {
    "guildId": "...",
    "track": {
      "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000,
      "isStream": false,
      "position": 0,
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    },
    "volume": 100,
    "paused": false,
    "voice": {
      "token": "...",
      "endpoint": "...",
      "sessionId": "...",
      "connected": true,

>       "ping": 10 // -1 if not connected

Reviewer 4: Should probably specify this is in milliseconds

    },
    "filters": { ... }
  }
]
```

### Get Player API
Returns the player for this guild in this session.
```
GET /v3/sessions/{sessionId}/players/{guildId}
Authorization: youshallnotpass
```

Response:
```json
{
  "guildId": "...",
  "track": {
    "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000,
    "isStream": false,
    "position": 0,
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,

>     "ping": 10 // -1 if not connected

Reviewer 4: Same again here (specify ms)

  },
  "filters": { ... }
}
```

### Update Player API
Updates the player for this guild in this specific.
```
PATCH /v3/sessions/{sessionId}/players/{guildId}?noReplace=true
Authorization: youshallnotpass
```

All fields are optional.

> You can provide eiter `trackData` or `identifier` or nothing, not both.

Reviewer 4: either

}
```

### Get Players API
Returns all players in this specific session.
```
GET /v3/sessions/{sessionId}/players

> Authorization: youshallnotpass

Reviewer 4: It may be worth removing this and instead detailing that such endpoints require an "Authorization" header containing the password specified in the application.yml, just to prevent any confusion in case folks take this a little too literally.

```

Response:
```json
[
  {
    "guildId": "...",
    "track": {
      "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000, // in ms
      "isStream": false,
      "position": 0, // in ms
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    },
    "volume": 100,
    "paused": false,
    "voice": {
      "token": "...",
      "endpoint": "...",
      "sessionId": "...",
      "connected": true,
      "ping": 10 // in ms & -1 if not connected
    },
    "filters": { ... }
  }
]
```

### Get Player API
Returns the player for this guild in this session.
```
GET /v3/sessions/{sessionId}/players/{guildId}

> Authorization: youshallnotpass

Reviewer 4: Same again here.

```

Response:
```json
{
  "guildId": "...",
  "track": {
    "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000, // in ms
    "isStream": false,
    "position": 0, // in ms
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  "volume": 100,
  "paused": false,
  "voice": {
    "token": "...",
    "endpoint": "...",
    "sessionId": "...",
    "connected": true,
    "ping": 10 // in ms & -1 if not connected
  },
  "filters": { ... }
}
```

### Update Player API
Updates the player for this guild in this specific.
```
PATCH /v3/sessions/{sessionId}/players/{guildId}?noReplace=true

> Authorization: youshallnotpass

Reviewer 4: See above.

  "rotation": {
    "rotationHz": 0 // The frequency of the audio rotating around the listener in Hz. 0.2 is similar to the example video above.
  },
  
  // Distortion effect. It can generate some pretty unique audio effects.
  "distortion": {
    "sinOffset": 0.0,
    "sinScale": 1.0,
    "cosOffset": 0.0,
    "cosScale": 1.0,
    "tanOffset": 0.0,
    "tanScale": 1.0,
    "offset": 0.0,
    "scale": 1.0
  } 
  
  // Mixes both channels (left and right), with a configurable factor on how much each channel affects the other.
  // With the defaults, both channels are kept independent from each other.
  // Setting all factors to 0.5 means both channels get the same audio.
  "channelMix": {
    "leftToLeft": 1.0,
    "leftToRight": 0.0,
    "rightToLeft": 0.0,
    "rightToRight": 1.0,
  }
  
  // Higher frequencies get suppressed, while lower frequencies pass through this filter, thus the name low pass.
  // Any smoothing values equal to, or less than 1.0 will disable the filter.
  "lowPass": {
    "smoothing": 20.0 // 1.0 < x
  }
}
```

### Destroy Player API
Destroys the player for this guild in this session.
```
DELETE /v3/sessions/{sessionId}/players/{guildId}

> Authorization: youshallnotpass

Reviewer 4: See above.

```

Response: 

> 204 - No Content

Reviewer 4: Perhaps 404 should be returned where a player doesn't exist?

### Update Session API
Updates the session with a resuming key and timeout.
```
PATCH /v3/sessions/{sessionId}

> Authorization: youshallnotpass

Reviewer 4: See above.

```

Request:
```json
{
    "resumingKey": "...",
    "timeout": 0 // in seconds
}
```

Response:

> 204 - No Content

Reviewer 4: 404 here as well if sessionId not found?

Reviewer: Yes, but that should not be documented here as it would apply to all endpoints with that path variable

### Track Loading API

> The REST api is used to resolve audio tracks for use with the `play` op.

Reviewer 4: is `play` op the correct terminology or is `play` endpoint meant instead?

Author: it refers to the "old" ws ops which it shouldn't

> `/loadtracks?identifier=dQw4w9WgXcQ` is deprecated and for removal in v4
```
GET /v3/loadtracks?identifier=dQw4w9WgXcQ

> Authorization: youshallnotpass

Reviewer 4: See above.

  "identifier": "dQw4w9WgXcQ",
  "isSeekable": true,
  "author": "RickAstleyVEVO",
  "length": 212000, // in ms
  "isStream": false,
  "position": 0, // in ms
  "title": "Rick Astley - Never Gonna Give You Up",
  "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
  "sourceName": "youtube"
}
```

> Decode multiple tracks into info their info

Reviewer 4: This reads weirdly.

> `/decodetracks` is deprecated and for removal in v4
```
POST /v3/decodetracks

> Authorization: youshallnotpass

Reviewer 4: See above.

    "info": {
      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000,
      "isStream": false,
      "position": 0,
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    }
  },
  ...
]
```
</details>

Response:
```json
[
  {
    "trackData": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000, // in ms
    "isStream": false,
    "position": 0, // in ms
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  ...
]
```
---

### Get list of plugins
Request information about the plugins running on Lavalink, if any.
> `/plugins` is deprecated and for removal in v4
```
GET /v3/plugins

> Authorization: youshallnotpass

Reviewer 4: See above.

]
```

### Get Lavalink version
Request Lavalink version,
> `/version` is deprecated and for removal in v4
```
GET /v3/version

> Authorization: youshallnotpass

Reviewer 4: See above.

```

Response:
```
v3.5
```

### RoutePlanner API

Additionally, there are a few REST endpoints for the ip rotation extension

#### Get RoutePlanner status
> `/routeplanner/status` is deprecated and for removal in v4
```
GET /v3/routeplanner/status

> Authorization: youshallnotpass

Reviewer 4: See above.

block.

#### Unmark a failed address
> `/routeplanner/free/address` is deprecated and for removal in v4
```
POST /v3/routeplanner/free/address
Host: localhost:8080

> Authorization: youshallnotpass

Reviewer 4: See above.

204 - No Content

#### Unmark all failed address
> `/routeplanner/free/all` is deprecated and for removal in v4
```
POST /v3/routeplanner/free/all
Host: localhost:8080

> Authorization: youshallnotpass

Reviewer 4: See above.

| tracks        | array of [Tracks](#track)              | All tracks which have been loaded                         | `TRACK_LOADED`, `PLAYLIST_LOADED`, `SEARCH_RESULT` |
| exception?    | [Exception](#exception-object) object  | The [Exception](#exception-object) this load failed with  | `LOAD_FAILED`                                      |

###### Load Result Type

| Load Result Type  | Description                                  |
|-------------------|----------------------------------------------|
| `TRACK_LOADED`    | A track has been loaded                      |
| `PLAYLIST_LOADED` | A playlist has been loaded                   |
| `SEARCH_RESULT`   | A search result has been loaded              |
| `NO_MATCHES`      | There has been no matches to your identifier |
| `LOAD_FAILED`     | Loading has failed                           |

###### Playlist Info

| Field         | Type   | Description                                                     |
|---------------|--------|-----------------------------------------------------------------|
| name          | string | The name of the loaded playlist                                 |

> | selectedTrack | int    | The selected track in this Playlist(-1 if no track is selected) |

Reviewer 2: ```suggestion
| selectedTrack | int    | The selected track in this Playlist (-1 if no track is selected) |
```

| exception?    | [Exception](#exception-object) object  | The [Exception](#exception-object) this load failed with  | `LOAD_FAILED`                                      |

##### Load Result Type

| Load Result Type  | Description                                  |
|-------------------|----------------------------------------------|
| `TRACK_LOADED`    | A track has been loaded                      |
| `PLAYLIST_LOADED` | A playlist has been loaded                   |
| `SEARCH_RESULT`   | A search result has been loaded              |
| `NO_MATCHES`      | There has been no matches to your identifier |
| `LOAD_FAILED`     | Loading has failed                           |

##### Playlist Info

| Field         | Type   | Description                                                      |
|---------------|--------|------------------------------------------------------------------|
| name          | string | The name of the loaded playlist                                  |
| selectedTrack | int    | The selected track in this Playlist (-1 if no track is selected) |

<details>
<summary>Track Loaded Example Payload</summary>

```yaml
{
  "loadType": "TRACK_LOADED",
  "tracks": [
    {

>       "encoded": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",

Reviewer 5: Same as my previous comment

  ]
}
```
</details>

<details>
<summary>Search Result Example Payload</summary>

```yaml
{
  "loadType": "SEARCH_RESULT",
  "tracks": [
    ...
  ]
}
```
</details>

<details>
<summary>No Matches Example Payload</summary>

```json
{
  "loadType": "NO_MATCHES"
}
```
</details>

<details>
<summary>Load Failed Example Payload</summary>

```json
{
  "loadType": "LOAD_FAILED",
  "exception": {
    "message": "The uploader has not made this video available in your country.",
    "severity": "COMMON"
  }
}
```
</details>

#### Track Searching
Lavalink supports searching via YouTube, YouTube Music, and Soundcloud. To search, you must prefix your identifier with `ytsearch:`, `ytmsearch:`,  or `scsearch:`respectively.

When a search prefix is used, the returned `loadType` will be `SEARCH_RESULT`. Note that, disabling the respective source managers renders these search prefixes redundant. Plugins may also implement prefixes to allow for more search engines.

---

#### Track Decoding

Decode a single track into its info

> > `/decodetrack?track=QAAAjJ...AAA==` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/decodetrack?track=QAAAjJ...AAA==` is deprecated and marked for removal in v4
```

  ]
}
```
</details>

Response:

##### Track Loading Result

| Field        | Type                                   | Description                                               | Valid Types                                        |
|--------------|----------------------------------------|-----------------------------------------------------------|----------------------------------------------------|
| loadType     | [LoadResultType](#load-result-type)    | The type of the result                                    | all                                                |
| playlistInfo | [Playlist Info](#playlist-info) object | Additional info if the the load type is `PLAYLIST_LOADED` | `PLAYLIST_LOADED`                                  |
| tracks       | array of [Tracks](#track)              | All tracks which have been loaded                         | `TRACK_LOADED`, `PLAYLIST_LOADED`, `SEARCH_RESULT` |

> | exception?   | [Exception](#exception-object) object  | The [Exception](#exception-object) this load failed with  | `LOAD_FAILED`                                      |

Reviewer 5: `exception` is present in the current response for all load result types - I'm not opposed to it but just wanted to mention it

example of the current response:
```json
{
    "loadType": "NO_MATCHES",
    "tracks": [],
    "playlistInfo": {
        "name": "",
        "selectedTrack": 0
    },
    "exception": null
}
```

Reviewer 6: Topi had broke the audio loader and had implemented a quick fix after I talked to him about it, he's not sure how the field omitting is going work quite yet.

Author: the json returned in https://github.com/freyacodes/Lavalink/pull/748 should match this documentation. So I'm gonna resolve this as that's not an issue with the proposal but rather the implemenation

| jvm            | string                             | The JVM version this Lavalink server runs on                    |
| lavaplayer     | string                             | The Lavaplayer version being used by this server                |
| sourceManagers | array of strings                   | The source manager names being used by this server              |
| plugins        | array of [Plugins](#plugin-object) | The plugin names being used by this server                      |

##### Version Object

| Field  | Type   | Description                               |
|--------|--------|-------------------------------------------|
| string | string | The full version of this Lavalink server  |
| major  | int    | The major version of this Lavalink server |
| minor  | int    | The minor version of this Lavalink server |
| patch  | int    | The patch version of this Lavalink server |

##### Git Object

| Field      | Type   | Description                             |
|------------|--------|-----------------------------------------|
| branch     | string | The branch of this Lavalink server      |
| commit     | string | The commit of this Lavalink server      |

> | commitTime | string | The commit time of this Lavalink server |

Reviewer 5: Is this meant to be a string or an int? asking because the Payload below is currently an int - I assume it just a timestamp and in turn an int

Author: this should be an int ty :)

##### Plugin Object

| Field   | Type   | Description               |
|---------|--------|---------------------------|
| name    | string | The name of the plugin    |
| version | string | The version of the plugin |

<details>
<summary>Example Payload</summary>

```json
{
  "version": {
    "string": "v3.7.0",
    "major": 3,
    "minor": 7,
    "patch": 0
  },
  "buildTime": 1664223916812,
  "git": {
    "branch": "master",
    "commit": "85c5ab5",
    "commitTime": 1664223916812
  },
  "jvm": "18.0.2.1",
  "lavaplayer": "1.3.98.4-original",

>   "sourceManagers": [

Reviewer 5: Are these the same as the `sourceName` returned in the TrackObject ?

Author: they are

  ]
}
```
</details>

<details>
<summary>Search Result Example Payload</summary>

```yaml
{
  "loadType": "SEARCH_RESULT",
  "tracks": [
    ...
  ]
}
```
</details>

<details>
<summary>No Matches Example Payload</summary>

```json
{
  "loadType": "NO_MATCHES"
}
```
</details>

<details>
<summary>Load Failed Example Payload</summary>

```json
{
  "loadType": "LOAD_FAILED",
  "exception": {
    "message": "The uploader has not made this video available in your country.",
    "severity": "COMMON"
  }
}
```
</details>

#### Track Searching
Lavalink supports searching via YouTube, YouTube Music, and Soundcloud. To search, you must prefix your identifier with `ytsearch:`, `ytmsearch:`,  or `scsearch:`respectively.

When a search prefix is used, the returned `loadType` will be `SEARCH_RESULT`. Note that, disabling the respective source managers renders these search prefixes redundant. Plugins may also implement prefixes to allow for more search engines.

---

#### Track Decoding

Decode a single track into its info, where `BASE64` is the encoded base64 data.

```

> GET /v3/decodetrack?track=BASE64

Reviewer 5: Wondering should track here also be deprecated in favour of encodedTrack? For consistency.

Author: not sure if I like camelCase as a query paramter `encoded-track` instead?

Reviewer 5: works for me - suggested camelCase for consistency and other endpoints using it (e.g noReplace)

Author: Oh yeah we use camelcase there 
Then it's probably fine I guess hm

  "sourceName": "youtube"
}
```
</details>

<details>
<summary>Response(Deprecated)</summary>

```json
{
  "track": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
  "info": {
    "identifier": "dQw4w9WgXcQ",
    "isSeekable": true,
    "author": "RickAstleyVEVO",
    "length": 212000,
    "isStream": false,
    "position": 0,
    "title": "Rick Astley - Never Gonna Give You Up",
    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  }
}
```
</details>

---

Decodes multiple tracks into their info

> > `/decodetracks` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/decodetracks` is deprecated and marked for removal in v4
```

    "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
    "sourceName": "youtube"
  },
  ...
]
```
</details>

<details>
<summary>Response(Deprecated)</summary>

```yaml
[
  {
    "track": "QAAAjQIAJVJpY2sgQXN0bGV5IC0gTmV2ZXIgR29ubmEgR2l2ZSBZb3UgVXAADlJpY2tBc3RsZXlWRVZPAAAAAAADPCAAC2RRdzR3OVdnWGNRAAEAK2h0dHBzOi8vd3d3LnlvdXR1YmUuY29tL3dhdGNoP3Y9ZFF3NHc5V2dYY1EAB3lvdXR1YmUAAAAAAAAAAA==",
    "info": {
      "identifier": "dQw4w9WgXcQ",
      "isSeekable": true,
      "author": "RickAstleyVEVO",
      "length": 212000,
      "isStream": false,
      "position": 0,
      "title": "Rick Astley - Never Gonna Give You Up",
      "uri": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
      "sourceName": "youtube"
    }
  },
  ...
]
```
</details>

---

#### Get Lavalink info
Request Lavalink information.
```
GET /v3/info
```

Response:

##### Info Response

| Field          | Type                               | Description                                        |
|----------------|------------------------------------|----------------------------------------------------|
| version        | [Version](#version-object) object  | The version of this Lavalink server                |

> | builtTime      | int                                | The time this Lavalink jar was built               |

Author: sure this is in milliseconds

]
```

---

#### Get Lavalink version(DEPRECATED)
Request Lavalink version.
> `/version` is deprecated and for removal in v4 use [`/info`](#get-lavalink-info) instead
```
GET /version
```

Response:
```
v3.6.0
```

---

### RoutePlanner API

Additionally, there are a few REST endpoints for the ip rotation extension

#### Get RoutePlanner status

> > `/routeplanner/status` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/routeplanner/status` is deprecated and marked for removal in v4
```

##### IP Block Type

| IP Block Type  | Description         |
|----------------|---------------------|
| `Inet4Address` | The ipv4 block type |
| `Inet6Address` | The ipv6 block type |

##### Failing Address Object

| Field            | Type   | Description                                              |
|------------------|--------|----------------------------------------------------------|
| address          | string | The failing address                                      |
| failingTimestamp | int    | The timestamp when the address failed                    |
| failingTime      | string | The timestamp when the address failed as a pretty string |

<details>
<summary>Example Payload</summary>

```json
{
  "class": "RotatingNanoIpRoutePlanner",
  "details": {
    "ipBlock": {
      "type": "Inet6Address",
      "size": "1208925819614629174706176"
    },
    "failingAddresses": [
      {
        "address": "/1.0.0.0",
        "failingTimestamp": 1573520707545,
        "failingTime": "Mon Nov 11 20:05:07 EST 2019"
      }
    ],
    "blockIndex": "0",
    "currentAddressIndex": "36792023813"
  }
}
```
</details>

---

#### Unmark a failed address

> > `/routeplanner/free/address` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/routeplanner/free/address` is deprecated and marked for removal in v4
```

```
POST /v3/routeplanner/free/address
```

Request:

| Field   | Type   | Description                                                                 |
|---------|--------|-----------------------------------------------------------------------------|
| address | string | The address to unmark as failed. This address must be in the same ip block. |

<details>
<summary>Example Payload</summary>

```json
{
  "address": "1.0.0.1"
}
```
</details>

Response:

204 - No Content

---

#### Unmark all failed address

> > `/routeplanner/free/all` is deprecated and for removal in v4

Reviewer 2: ```suggestion
> `/routeplanner/free/all` is deprecated and marked for removal in v4
```

#### Get Lavalink version (DEPRECATED)
Request Lavalink version.
> `/version` is deprecated and marked for removal in v4, use [`/info`](#get-lavalink-info) instead
```
GET /version
```

Response:
```
v3.7.0
```

---

### RoutePlanner API

Additionally, there are a few REST endpoints for the ip rotation extension

#### Get RoutePlanner status
> `/routeplanner/status` is deprecated and marked for removal in v4
```
GET /v3/routeplanner/status
```

Response:

| Field   | Type                                        | Description                                                           |
|---------|---------------------------------------------|-----------------------------------------------------------------------|
| class   | ?[Route Planner Type](#route-planner-types) | The name of the RoutePlanner implementation being used by this server |
| details | ?[Details](#details-object) object          | The status details of the RoutePlanner                                |

##### Route Planner Types

| Route Planner Type           | Description |
|------------------------------|-------------|
| `RotatingIpRoutePlanner`     | ...         |
| `NanoIpRoutePlanner`         | ...         |

> | `RotatingNanoIpRoutePlanner` | ...         |

Author: Those 3 fields need a description but I have no idea

]
```

---

#### Get Lavalink version (DEPRECATED)
Request Lavalink version.
> `/version` is deprecated and marked for removal in v4, use [`/info`](#get-lavalink-info) instead
```

> GET /version

Reviewer 5: Could we keep a version-agnostic endpoint to query the node version?

The purpose of this would be to handle multiple versions of a node within a library.

This way I for example would be able to know if a node is v3 or v4 etc before fully connecting to it (and guessing the endpoints) and be able to support multiple versions more easily.

While yes I could try to call `/version` and `v3/info` and `v4/info` and try a different one if it errors or returns nothing - ideally I would query `/version` first to know if it is; say 3.6.0, 3.7.0 or 4.0.0 and call the correct endpoints from there.

For graceful compatibility with 3.6.0 and lower this should remain as `/version` and return nothing but the version number.

Reviewer: Good point, but at the same time I don't like the current response

Perhaps it should just return the major version and nothing else. We can include in the documentation that if it prefixed with `v` then v3 should be assumed.

Reviewer 5: Was hoping to have a version semantic output - just for more easily able to get the version (minors) as a one-op.

The way I see it is if you only return `{version}` as a text output prior to 3.7, `v3` for <4 and `v4` <5 I would then need to do a second follow-up call via `/v3/info` and `/v4/info` to know the full version name.

While that is not a massive issue it does feel unnecessary (making 2 calls in 3.7+) when keeping it as it is would make it doable in 1 call.

Reviewer 5: Just for context Right now with 3.5.1 and 3.6.0 both returns semver responses which is ideal imo - if we ensure to use semver (i.e "3.7.0" and not "v3.7.0" as proposed here), no special handling would be required on lib sides to allow them to parse the version with the regex expression in semver.org - and as long as the git tags remain semver compatible no further changes will ever be needed here.

Author: I undeprecated `/version` 👍🏻

```

Response:
```
v3.7.0
```

---

### RoutePlanner API

> Additionally, there are a few REST endpoints for the ip rotation extension

Reviewer 2: ```suggestion
Additionally, there are a few REST endpoints for the ip rotation extension.
```

#### Get RoutePlanner status
> `/routeplanner/status` is deprecated and marked for removal in v4
```
GET /v3/routeplanner/status
```

Response:

| Field   | Type                                        | Description                                                           |
|---------|---------------------------------------------|-----------------------------------------------------------------------|
| class   | ?[Route Planner Type](#route-planner-types) | The name of the RoutePlanner implementation being used by this server |
| details | ?[Details](#details-object) object          | The status details of the RoutePlanner                                |

##### Route Planner Types

> | Route Planner Type           | Description |
> |------------------------------|-------------|
> | `RotatingIpRoutePlanner`     | ...         |
> | `NanoIpRoutePlanner`         | ...         |
> | `RotatingNanoIpRoutePlanner` | ...         |

Reviewer 2: ```suggestion
| Route Planner Type           | Description                                                                                                                  |
|------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| `RotatingIpRoutePlanner`     | IP address used is switched on ban. Recommended for IPv4 blocks or IPv6 blocks smaller than a /64.                            |
| `NanoIpRoutePlanner`         | IP address used is switched on clock update. Use with at least 1 /64 IPv6 block.                                             |
| `RotatingNanoIpRoutePlanner` | IP address used is switched on clock update, rotates to a different /64 block on ban. Use with at least 2x /64 IPv6 blocks.  |
```

---

### Resuming Lavalink Sessions

What happens after your client disconnects is dependent on whether the session has been configured for resuming.

* If resuming is disabled all voice connections are closed immediately.
* If resuming is enabled all music will continue playing. You will then be able to resume your session, allowing you to control the players again.

> To enable resuming, you must call the [Update Session](#update-session) endpoint with the `resumingKey` and `timeout`

Reviewer 2: ```suggestion
To enable resuming, you must call the [Update Session](#update-session) endpoint with the `resumingKey` and `timeout`.
```

<summary>Configure Resuming OP(DEPRECATED)</summary>

To enable resuming, you must send a `configureResuming` message.

* `key` is the string you will need to send when resuming the session. Set to null to disable resuming altogether. Defaults to null.
* `timeout` is the number of seconds after disconnecting before the session is closed anyway. This is useful for avoiding accidental leaks. Defaults to `60` (seconds).

```json
{
  "op": "configureResuming",
  "key": "myResumeKey",
  "timeout": 60
}
```
</details>

To resume a session, specify the resume key in your WebSocket handshake request headers:

```
Resume-Key: The resume key of the session you want to resume.
```

You can tell if your session was resumed by looking at the handshake response header `Session-Resumed` which is either `true` or `false`.
```
Session-Resumed: true
```

In case your websocket library doesn't support reading headers you can listen for the [ready op](#ready-op) and check the `resumed` field.

When a session is paused, any events that would normally have been sent is queued up. When the session is resumed, this
queue is then emptied and the events are then replayed. 

---

### Special notes

> * When your shard's main WS connection dies, so does all your lavalink audio connections.

Reviewer 2: ```suggestion
* When your shard's main WS connection dies, so does all your Lavalink audio connections.
```
